src/WebJobs.Extensions.DocumentDB/Bindings/DocumentDBAttributeBindingProvider.cs

```diff
@@ -0,0 +1,131 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Globalization;
+using System.Net;
+using System.Reflection;
+using System.Runtime.ExceptionServices;
+using System.Threading.Tasks;
+using Microsoft.Azure.Documents;
+using Microsoft.Azure.Documents.Client;
+using Microsoft.Azure.WebJobs.Host;
+using Microsoft.Azure.WebJobs.Host.Bindings;
+
+namespace Microsoft.Azure.WebJobs.Extensions.DocumentDB
+{
+    internal class DocumentDBAttributeBindingProvider : IBindingProvider
+    {
+        private JobHostConfiguration _jobHostConfig;
+        private DocumentDBConfiguration _docDBConfig;
+
+        public DocumentDBAttributeBindingProvider(JobHostConfiguration config, DocumentDBConfiguration documentDBConfig)
+        {
+            _jobHostConfig = config;
+            _docDBConfig = documentDBConfig;
+        }
+
+        public async Task<IBinding> TryCreateAsync(BindingProviderContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException("context");
+            }
+
+            ParameterInfo parameter = context.Parameter;
+            DocumentDBAttribute attribute = parameter.GetCustomAttribute<DocumentDBAttribute>(inherit: false);
+            if (attribute == null)
+            {
+                return null;
+            }
+
+            if (_docDBConfig.EndpointUri == null)
+            {
+                throw new InvalidOperationException(
+                    string.Format(CultureInfo.CurrentCulture,
+                    "The DocumentDB Endpoint Uri must be set either via a '{0}' app setting or directly in code via DocumentDBConfiguration.EndpointUri.",
+                    DocumentDBConfiguration.AzureWebJobsDocumentDBUriName));
+            }
+
+            if (string.IsNullOrEmpty(_docDBConfig.AuthKey))
+            {
+                throw new InvalidOperationException(
+                    string.Format(CultureInfo.CurrentCulture,
+                    "The DocumentDB Authentication Key must be set either via a '{0}' app setting or directly in code via DocumentDBConfiguration.AuthKey.",
+                    DocumentDBConfiguration.AzureWebJobsDocumentDBKeyName));
+            }
+
+            DocumentDBContext documentDBContext = CreateContext(_docDBConfig, attribute, _jobHostConfig.NameResolver);
+
+            await CreateIfNotExists(attribute, documentDBContext);
```

Comment:
That was my intention, yes. DocDB team recommended doing this during a setup phase. The other options could result in a slower initial invocation. I wasn't sure if that was a good plan either.

I originally had a check for 404 built into the AsyncCollector that would then create the db and collection if they weren't there but that seemed messy and I didn't like the idea of doing that in every other (future) binding as well.

Other bindings (Queue, Table) do this in the call to BindAsync which I believe happens for every invocation. That seems heavy too. 

I vote we keep it like this and come back to it if there's a better pattern... thoughts?

---
In reply to: [56258300](https://github.com/Azure/azure-webjobs-sdk-extensions/pull/47#discussion_r56258300) [](ancestors = 56258300)